python/dgl/contrib/dis_kvstore.py

```diff
@@ -220,195 +363,320 @@ def _push_handler(self, name, ID, data):
+        of each element in a target data tensor.
+
+        Parameters
+        ----------
+        name : str
+            data name
+        partition_book : list or tensor (mx.ndarray or torch.tensor)
+            A book that maps global ID to target server ID.
+        """
+        assert len(name) > 0, 'name cannot be empty.'
+        assert len(partition_book) > 0, 'partition_book cannot be empty.'
+
+        if isinstance(partition_book, list):
+            self._data_store[name+'-part-'] = F.tensor(partition_book)
+        else:
+            self._data_store[name+'-part-'] = partition_book
+
+        self._has_data.add(name+'-part-')
+
+
     def connect(self):
-        """Connect to all KVServer nodes
+        """Connect to all the KVServer nodes
         """
         for ID, addr in self._server_namebook.items():
             server_ip, server_port = addr.split(':')
             _add_receiver_addr(self._sender, server_ip, int(server_port), ID)
         _sender_connect(self._sender)
-        client_ip, client_port = self._addr.split(':')
-        _receiver_wait(self._receiver, client_ip, int(client_port), self._server_count)
 
-    def init_data(self, name, server_id, shape, init_type='zero', low=0.0, high=0.0):
-        """Initialize kvstore tensor
+        self._addr = self._get_local_addr()
+        client_ip, client_port = self._addr.split(':')
 
-        we hack the msg format here: msg.id store the shape of target tensor,
-        msg.data has two row, and the first row is the init_type, 
-        [0, 0] means 'zero' and [1,1] means 'uniform'. 
-        The second row is the min & max threshold.
+        # find local server nodes
+        for ID, addr in self._server_namebook.items():
+            server_ip, server_port = addr.split(':')
+            if client_ip == server_ip or server_ip == '127.0.0.1':
+                self._local_server_id.add(ID)
 
-        Parameters
-        ----------
-        name : str
-            data name
-        server_id : int
-            target server id
-        shape : list of int
-            shape of tensor
-        init_type : str
-            initialize method, including 'zero' and 'uniform'
-        low : float
-            min threshold, if use 'uniform'
-        high : float
-            max threshold, if use 'uniform'
-        """
-        tensor_shape = F.tensor(shape)
-        init_type = 0.0 if init_type == 'zero' else 1.0
-        threshold = F.tensor([[init_type, init_type], [low, high]])
+        # send addr to server nodes
         msg = KVStoreMsg(
-            type=KVMsgType.INIT,
-            rank=self._client_id,
-            name=name,
-            id=tensor_shape,
-            data=threshold)
-        _send_kv_msg(self._sender, msg, server_id)
-        
-    def push(self, name, server_id, id_tensor, data_tensor):
-        """Push sparse message to target KVServer.
+            type=KVMsgType.IP_ID,
+            rank=0,
```

Comment:
what does `rank` mean here?

Reply:
`rank` means the sender's ID. If the ID isn't been assigned, it could be 0 by default.